lib/interface-types/src/interpreter/instructions/memory_to_string.rs

```diff
@@ -1,53 +1,57 @@
         move |runtime| -> _ {
-            match runtime.stack.pop(2) {
-                Some(inputs) => match runtime.wasm_instance.memory(0) {
-                    Some(memory) => {
-                        let length = i32::try_from(&inputs[0])? as usize;
-                        let pointer = i32::try_from(&inputs[1])? as usize;
-                        let memory_view = memory.view();
+            let inputs = runtime.stack.pop(2).ok_or_else(|| {
+                InstructionError::new(
+                    instruction,
+                    InstructionErrorKind::StackIsTooSmall { needed: 2 },
+                )
+            })?;
 
-                        if memory_view.len() < pointer + length {
-                            return Err(format!(
-                                "`{}` failed because it has to read out of the memory bounds (index {} > memory length {}).",
-                                instruction_name,
-                                pointer + length,
-                                memory_view.len()
-                            ));
-                        }
+            let memory_index: u32 = 0;
 
-                        let data: Vec<u8> = (&memory_view[pointer..pointer + length])
-                            .iter()
-                            .map(Cell::get)
-                            .collect();
+            let memory = runtime
+                .wasm_instance
+                .memory(memory_index as usize)
+                .ok_or_else(|| {
+                    InstructionError::new(
+                        instruction,
+                        InstructionErrorKind::MemoryIsMissing { memory_index },
+                    )
+                })?;
 
-                        match String::from_utf8(data) {
-                            Ok(string) => {
-                                runtime.stack.push(InterfaceValue::String(string));
+            let length = to_native::<i32>(&inputs[0], instruction)? as usize;
+            let pointer = to_native::<i32>(&inputs[1], instruction)? as usize;
+            let memory_view = memory.view();
 
-                                Ok(())
-                            }
-                            Err(utf8_error) => Err(format!(
-                                "`{}` failed because the read string isn't UTF-8 valid ({}).",
-                                instruction_name,
-                                utf8_error,
-                            ))
-                        }
-                    }
-                    None => Err(format!(
-                        "`{}` failed because there is no memory to read.",
-                        instruction_name
-                    ))
-                }
-                None => Err(format!(
-                    "`{}` failed because there is not enough data on the stack (needs 2).",
-                    instruction_name,
-                ))
+            if memory_view.len() < pointer + length {
```

Comment:
I think this may be affected by the same bug I recently fixed in WasmPtr,

Given a String, "hi" at address 0xFFFE the memory looks like

0xFFFE: 'h'
0xFFFF: 'i'

ptr (0xFFFE) + len (2) = 0x10000, failing this bounds check.  The fix, in general, is annoying when you handle length = 0 and zero sized types.

See https://github.com/wasmerio/wasmer/pull/1272 for my fix.

Better to have false positives than false negatives in bounds-checking code, but leaving a comment here as a heads up.  It might make sense to do a full pass for these types of bugs and fix them in a separate PR if you think it might clutter this one.

Reply:
To clarify more, we want to make sure out of bounds pointers are never derefed, so ptr = 0x10000 and length = 0 will not be caught by the naive solution of changing `<` to `<=` here.

Reply:
I'm not sure I understand. I've pushed a new commit to use inclusive an range, and consequently a special condition to handle empty string, but I'm not sure there is a bug right now. The `test_memory_to_string` already reads at bounds.